### PR TITLE
cli: add inspect subcommand

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

There are three features that would be useful to snapcraft CLI developers that are not supported today:

  1. A quick-look summary of the current project state.
  2. A way to determine which lifecycle step of which part was most recently completed.
  3. A way to determine which part provided a specific file within the staging or priming area.

This PR resolves [LP: #1759700](https://bugs.launchpad.net/snapcraft/+bug/1759700) by adding support for all three of these use-cases by introducing a new, hidden subcommand called `inspect`.

  1. `snapcraft inspect` will print a quick summary of the current project state.
  2. `snapcraft inspect --latest-step` will print the part, step, and timestamp of the latest step that was run.
  3. `snapcraft inspect --provides <path>` will print the part that provided the path.

Default to human-readable output, but also support a `--json` to print json for programmatic consumption.

### CODE REVIEW

File: snapcraft/cli/parts.py [hunk 1]
```diff
@@ -68,3 +76,178 @@ def search(ctx, query):
...
+
+    By default, this command will print a tabulated summary of the current state of the
+    project.
+    """
+    count = 0
+    for option in (provides, latest_step):
+        if option:
+            count += 1
+
+    if count > 1:
```
Comment (sergiusens) [inline snapcraft/cli/parts.py]: I wonder if it is worth expanding click like in this last comment on this issue https://github.com/pallets/click/issues/257
Totally not a requirement for this PR, just a thought.

**Comment by kyrofa** [comment]
> There seems to be a problem when running on a not yet run project

Whoops, sorry, that was silly. Fixed.

> For the json output at least we would need to get the working directory for that step too.

Sure, also fixed.

File: snapcraft/cli/parts.py (hunk 1 above)
Comment (kyrofa) [reply snapcraft/cli/parts.py]: I think this is simple enough not to warrant more code, but if we find we need this functionality elsewhere, it may be worth doing.

File: snapcraft/internal/errors.py [hunk 2]
```diff
@@ -185,6 +185,18 @@ class SnapcraftPartMissingError(SnapcraftError):
     )
 
 
+class NoSuchFileError(SnapcraftError):
```
Comment (sergiusens) [inline snapcraft/internal/errors.py]: the error message seems to specific for this class' name. How about `NoSuchFileInPartError`.
These messages as with ToolError are getting a bit too un-namespaced btw. We should fix that.

File: snapcraft/internal/errors.py [hunk 3]
```diff
@@ -632,3 +644,34 @@ class InvalidMountinfoFormat(SnapcraftError):
...
+        "Failed to find part that provides path: {path!r} is not in the "
+        "staging or priming area.\n"
+        "Ensure the path is in the staging or priming area and try again."
+    )
+
+    def __init__(self, path):
+        super().__init__(path=path)
+
+
+class UntrackedFileError(SnapcraftInspectError):
```
Comment (sergiusens) [inline snapcraft/internal/errors.py]: this is specific to parts too

File: snapcraft/internal/errors.py [hunk 4]
```diff
@@ -632,3 +644,34 @@ class InvalidMountinfoFormat(SnapcraftError):
...
+
+    fmt = "No known parts provided {path!r}. It may have been provided by a scriptlet."
+
+    def __init__(self, path):
+        super().__init__(path=path)
+
+
+class NoStepsRunError(SnapcraftInspectError):
```
Comment (sergiusens) [inline snapcraft/internal/errors.py]: ok, I see the pattern, we should prepend SnapcraftInspect to the class names (or put them in their own inspect package with most of the logic in the cli

File: snapcraft/internal/pluginhandler/__init__.py [hunk 5]
```diff
@@ -126,24 +126,27 @@ def __init__(
 
     def get_pull_state(self) -> states.PullState:
         if not self._pull_state:
-            self._pull_state = states.get_state(self.plugin.statedir, steps.PULL)
+            self._pull_state = cast(states.PullState, self.get_state(steps.PULL))
```
Comment (sergiusens) [inline snapcraft/internal/pluginhandler/__init__.py]: why is this necessary?

Comment (kyrofa) [reply snapcraft/internal/pluginhandler/__init__.py]: Because this PR adds static checks to the return type of `get_state`. This now means casting if it's a subclass.

File: snapcraft/cli/parts.py [hunk 6]
```diff
@@ -68,3 +76,188 @@ def search(ctx, query):
...
+                latest_step = step
+                latest_timestamp = timestamp
+
+    if not latest_part:
+        raise errors.NoStepsRunError()
+
+    return (latest_part, latest_step, latest_timestamp)
+
+
+def _lifecycle_status(config):
```
[the remark this answers is not in the document]
Comment (kyrofa) [reply snapcraft/cli/parts.py]: Done.
